pkg/app/app_template.go

```diff
@@ -26,20 +26,40 @@ func (a *App) template(dirPath string) exec.CmdRunResult {
 	var result exec.CmdRunResult
 	var isStream bool
 
+	kubernetesVersion, err := a.compInfo.KubernetesVersion(a.app.Spec.ServiceAccountName, a.app.Spec.Cluster, &a.app.ObjectMeta)
+	if err != nil {
+		result.AttachErrorf("%s", fmt.Errorf("Unable to get kubernetes version before templating"))
+		return result
+	}
+
+	kappControllerVersion, err := a.compInfo.KappControllerVersion()
+	if err != nil {
+		result.AttachErrorf("%s", fmt.Errorf("Unable to get kapp-controller version before templating"))
+		return result
+	}
+
+	k8sAPIs, err := a.compInfo.KubernetesAPIs()
+	if err != nil {
+		result.AttachErrorf("%s", fmt.Errorf("Unable to list all server apigroups/version before templating"))
+		return result
+	}
+
```

Comment:
perf: we should not have to retrieve any of this info if it's not being consumed by values.

Reply:
Hmm, that's an interesting ask. So the reason it's like it is currently is:

1. I wanted to call each of these APIs only once before every templating step. Mainly because we have no good way to cache the k8s version on an external cluster, nor the k8s apis. We could do as Joe suggested with a one minute cache. 
2. The calls would have to get moved down into each factory individually (i.e duplicate the information gathering of something that ought not to change between templating steps.

Reply:
I went ahead and memoized the calls, I added a test which validates the calls are made made only once with multiple template steps. pkg/app/app_template_test.go